Upgrade to Wagtail 2.12.3

cwdavies opened the pull request:
In Wagtail 2.12 stream_data on StreamField values is deprecated

https://docs.wagtail.io/en/stable/releases/2.12.html


## Additions

- Add raw_data to export_mega_menu management command
- Add raw_data to import_mega_menu management command
- @willbarton worked on a fix for the wagtailembeds migration failure issue

## Changes

I took the approach of using `raw_data` as `steam_data` is deprecated. In `cfgov/v1/util/migrations.py`, `get_streamfield_data` and `set_streamfield_data` are modified to use `raw_data`.

As a result of the above change I fixed a TypeError: Object of type 'RawDataView' is not JSON serializable


## How to test this PR

- Verify there are no failures running wagtail migrations and check generated indexes

1. tox -e unittest-future
2. python cfgov/manage.py migrate
3. python cfgov/manage.py runscript export_enforcement_actions
4. python cfgov/manage.py runscript update_oaa_status
5. python cfgov/manage.py runscript export_ask_data
6. python cfgov/manage.py runscript fix_links_in_tableblocks
7. python cfgov/manage.py update_chart_block_dates --snapshot_file

1. cfgov/manage.py export_mega_menu en output.json
2. cfgov/manage.py export_mega_menu es output.json OR
3. cfgov/manage.py export_mega_menu en (writes to stdout)

4. cfgov/manage.py import_mega_menu en import.json
5. cfgov/manage.py import_mega_menu es import.json OR
6. cfgov/manage.py import_mega_menu en (reads from stdin) 

cat new.json | cfgov/manage.py import_mega_menu en

Current menu:

curl -k https://github.local/raw/gist/chosak/4600af0d52fbe588a634ea6fee54f942/raw/5af1c1428f63ddfd7f6fc8ba9
fa115b0ac17235a/current.json | cfgov/manage.py import_mega_menu en

New menu:

curl -k https://github.local/raw/gist/chosak/fdeab64f382ca0d17568cdbd717862d8/raw/9b2e8fb73062ec631c609d69c1e9ee49261b6459/new.json | cfgov/manage.py import_mega_menu en

After importing a new mega menu JSON file, verify changes to the mega menu are visible in http://localhost:8000/


## Notes and todos

- Send email notification before merging this PR


## Checklist

- [x] PR has an informative and human-readable title
  - PR titles are used to generate the change log in [releases](../../releases); good ones make that easier to scan.
  - Consider prefixing, e.g., "Mega Menu: fix layout bug", or "Docs: Update Docker installation instructions".
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code follows the standards laid out in the [CFPB development guidelines](https://github.com/cfpb/development)
- [x] Future todos are captured in comments and/or tickets
- [ ] Project documentation has been updated, potentially one or more of:
  - [This repo’s docs](https://cfpb.github.io/consumerfinance.gov/) (edit the files in the `/docs` folder) – for basic, close-to-the-code docs on working with this repo
  - CFGOV/platform wiki on GHE – for internal CFPB developer guidance
  - CFPB/hubcap wiki on GHE – for internal CFPB design and content guidance

#### Browser testing

Visually tested in the following supported browsers:
- [ ] Firefox
- [ ] Chrome
- [ ] Safari
- [ ] Edge 18 (the last Edge prior to it switching to Chromium)
- [ ] Internet Explorer 11 and 8 (via emulation in 11's dev tools)
- [ ] Safari on iOS
- [ ] Chrome on Android

#### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Does not introduce new lint warnings
- [ ] Flexible from small to large screens

willbarton commented:
> I took the approach of using raw_data as steam_data is deprecated. I renamed get_stream_data and set_stream_data in cfgov/v1/util/migrations.py, to get_data and set_data, and modified them to use raw_data.

Hmm, one thing I think I'd like to see changed here is the naming. I think `set_data` and `get_data` aren't quite as descriptive as I think we'd want them to do. They're only really useful on streamfields, which was *mostly* implied in `stream_data`.

Could we make these `get_streamfield_data` and `set_streamfield_data`?

willbarton commented on cfgov/ask_cfpb/scripts/export_ask_data.py:
```diff
@@ -61,7 +63,10 @@ def assemble_output():
         output['RelatedResource'] = page['related_resource__title']
         output['Question'] = page['question'].replace('\x81', '')
-        answer_streamfield = page['answer_content'].stream_data
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
Since you've settled on just pinning Wagtail to 2.12.3 in `requirements/wagtail.txt`, this doesn't seem necessary.

cwdavies replied:
Agreed.

willbarton commented on cfgov/data_research/research_conference.py:
```diff
@@ -23,7 +24,11 @@ def get_conference_details_from_page(page_id):
 
     if hasattr(page, 'content'):
-        for block in page.content.stream_data:
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
Same.

cwdavies replied:
Done

willbarton commented on cfgov/mega_menu/management/commands/export_mega_menu.py:
```diff
@@ -20,8 +22,8 @@ def add_arguments(self, parser):
-            options['filename'],
-            indent=4
-        )
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
Same.

cwdavies replied:
Done

willbarton commented on cfgov/scripts/fix_links_in_tableblocks.py:
```diff
@@ -15,13 +16,16 @@ def get_tableblocks(page):
     """
     try:
-        stream_data = page.specific.content.stream_data
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
Same.

cwdavies replied:
Done

willbarton commented on cfgov/v1/management/commands/update_chart_block_dates.py:
```diff
@@ -45,9 +47,13 @@ def update_chart_blocks(self, date_published, last_updated, markets):
         """ Update date_published on all chart blocks """
 
         for page in BrowsePage.objects.all():
+            if wagtail.VERSION < (2, 12):  # pragma: no cover
```
Same.

cwdavies replied:
Done

willbarton commented on cfgov/v1/management/commands/update_data_snapshot_values.py:
```diff
@@ -37,10 +39,13 @@ def get_data_snapshots(self):
-            if stream_data and stream_data[0]['type'] == 'data_snapshot':
-                stream_data[0]['value']['page'] = page.pk
-                snapshots.append(stream_data)
+            if wagtail.VERSION < (2, 12):  # pragma: no cover
```
Same.

cwdavies replied:
Done

willbarton commented on cfgov/v1/util/migrations.py:
```diff
@@ -68,30 +69,33 @@ def is_page(page_or_revision):
     if is_page(page_or_revision):
         field = getattr(page_or_revision, field_name)
-        return field.stream_data
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
Same.

cwdavies replied:
Done

cwdavies commented:
> > I took the approach of using raw_data as steam_data is deprecated. I renamed get_stream_data and set_stream_data in cfgov/v1/util/migrations.py, to get_data and set_data, and modified them to use raw_data.
> 
> Hmm, one thing I think I'd like to see changed here is the naming. I think `set_data` and `get_data` aren't quite as descriptive as I think we'd want them to do. They're only really useful on streamfields, which was _mostly_ implied in `stream_data`.
> 
> Could we make these `get_streamfield_data` and `set_streamfield_data`?

That's a great suggestion, so I have renamed to be more descriptive to get, set, and migrate streamfield data.

willbarton commented on cfgov/mega_menu/management/commands/export_mega_menu.py:
```diff
@@ -21,7 +21,5 @@ def add_arguments(self, parser):
-            menu.submenus.stream_data,
-            options['filename'],
-            indent=4
+            list(menu.submenus.raw_data), options['filename'], indent=4
```
Super minor style thing:

If the arguments are going to be on a new line from the function call, they should each have their own line, like they did prior.

cwdavies replied:
Good catch and I made this change. Usually lint picks up changes needed for styling.

willbarton commented on cfgov/mega_menu/tests/management/commands/test_import_mega_menu.py:
```diff
@@ -40,4 +42,7 @@ def test_import(self):
 
         menu = Menu.objects.get(language='en')
-        self.assertEqual(menu.submenus.stream_data, submenus)
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
Oops, here's another version fence I missed!

cwdavies replied:
Removed wagtail.VERSION < from tests

willbarton commented on cfgov/v1/tests/util/test_migrations.py:
```diff
@@ -36,53 +36,59 @@ def test_is_page_revision(self):
+        set_streamfield_data(self.page, 'body', new_data)
 
-        self.assertEqual(self.page.body.stream_data[0]['value'], 'new text')
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
One more version fence here too.

cwdavies replied:
Removed wagtail.VERSION < from tests

willbarton commented on cfgov/v1/tests/util/test_migrations.py:
```diff
@@ -36,53 +36,59 @@ def test_is_page_revision(self):
-        self.assertEqual(
-            self.revision.as_page_object().body.stream_data[0]['value'],
-            'new text')
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
Oops, and here!

cwdavies replied:
Removed wagtail.VERSION < from tests

willbarton commented on cfgov/v1/tests/util/test_migrations.py:
```diff
@@ -91,35 +97,38 @@ def test_migrate_stream_field_page(self):
         mapper.assert_called_with(self.page, 'some text')
-        self.assertEqual(self.page.body.stream_data[0]['value'],
-                         'new text')
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
And again!

cwdavies replied:
Removed wagtail.VERSION < from tests

willbarton commented on cfgov/v1/tests/util/test_migrations.py:
```diff
@@ -91,35 +97,38 @@ def test_migrate_stream_field_page(self):
-        self.assertEqual(
-            self.revision.as_page_object().body.stream_data[0]['value'],
-            'new text')
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
And here!

cwdavies replied:
Removed wagtail.VERSION < from tests

willbarton commented on cfgov/v1/tests/util/test_migrations.py:
```diff
@@ -189,41 +198,45 @@ def setUp(self):
         self.block = TestStreamBlock()
-        self.value = self.block.to_python(self.original_stream_data)
+        self.value = self.block.to_python(self.original_data)
+        if wagtail.VERSION < (2, 12):  # pragma: no cover
```
And here!

cwdavies replied:
Removed wagtail.VERSION < from tests

willbarton reviewed (commented):
A few more super-minor things, and then I'll think we'll be about ready to go!

willbarton commented:
Do you see an error on a production database with the migration of wagtailembeds?

```
  Applying wagtailembeds.0008_allow_long_urls...Traceback (most recent call last):
  File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.ProgrammingError: constraint "idx_17439_wagtailembeds_embed_url_37a13a49926a4846_uniq" of relation "wagtailembeds_embed" does not exist
```

cwdavies commented:
> Do you see an error on a production database with the migration of wagtailembeds?
> 
> ```
>   Applying wagtailembeds.0008_allow_long_urls...Traceback (most recent call last):
>   File "/opt/rh/rh-python36/root/usr/lib/python3.6/site-packages/django/db/backends/utils.py", line 84, in _execute
>     return self.cursor.execute(sql, params)
> psycopg2.ProgrammingError: constraint "idx_17439_wagtailembeds_embed_url_37a13a49926a4846_uniq" of relation "wagtailembeds_embed" does not exist
> ```

@willbarton Yes. I see the exact same error with wagtailembeds when I ran the refresh-data.sh script.

willbarton commented:
I'm digging into the `wagtailembeds` migration failure. That's a major blocker for right now.

We should also look at bumping our far-outdated django-treebeard requirement.

higs4281 reviewed (approved):
Generated indexes look good now in test deploy.
